src/Data/Sequence/Circular.hs

```diff
@@ -29,114 +26,48 @@ import qualified Data.Sequence as Seq
 type Limit = Int
 type Size  = Int
 
-data CircularSeq a = CSeq !Limit {-# UNPACK #-} !Size !(Seq.Seq a)
+-- | The index to the head of the vector representing the sequence.
+-- This \"pointer\" starts at the end of the vector and \"wraps\" when we
+-- reach 0. This ensure we can implement 'toList' efficiently by avoiding
+-- needless calls to 'reverse'.
+type Head  = Int
 
-instance Show a => Show (CircularSeq a) where
-   show (CSeq _ _ x) = show x
-
-instance Eq a => Eq (CircularSeq a) where
-    (CSeq _ _ x) == (CSeq _ _ y) = x == y
-
-instance Ord a => Ord (CircularSeq a) where
-    (CSeq _ _ x) <= (CSeq _ _ y) = x <= y
+data CircularSeq a =
+    CSeq                 Limit
+          {-# UNPACK #-} Size
+          {-# UNPACK #-} Head
+          (V.Vector a)
 
-instance Functor CircularSeq where
-  fmap f (CSeq l v x) = CSeq l v (fmap f x)
-
-instance Foldable CircularSeq where
-  foldMap f (CSeq _ _ x) = foldMap f x
+instance Show a => Show (CircularSeq a) where
+   show = show . toList
 
 {------------------------------------------------------------------------------
   Main API
 ------------------------------------------------------------------------------}
 
+toList :: CircularSeq a -> [a]
```

Comment:
Could you document the ordering of the elements in the list? I.e. the most recently pushed element is at the end while the oldest is at the beginning

Reply:
ps. note that the ordering is the opposite of what you are describing -- due to the fact we expose only a "cons" operation (i.e. <|), elements are added in front of the sequence, i.e:

```haskell
> import qualified Data.Sequence.Circular as CSeq
> let c = CSeq.empty 3
> 1 CSeq.<| c
[1]
> 2 CSeq.<| it
[2,1]
```

We could expose a `|>` function as before, but I have removed it as it's not needed for our use case (i.e. the block cache).